<p>В языке Java все исключения являются объектами и могут быть вызваны не только системой,
но и создаваться самим разработчиком.
Исключения делятся на несколько классов, которые имеют общего предка — класс <code>Throwable</code>.
Его потомками являются подклассы <code>Exception</code> и <code>Error</code>.
Исключения (<strong><code>Exceptions</code></strong>) являются результатом проблем в программе, которые в принципе решаемы и предсказуемы.
Например, произошло деление на ноль в целых числах.</p>
<p>Ошибки (<strong><code>Errors</code></strong>) представляют собой более серьёзные проблемы, которые, согласно спецификации <code>Java</code>,
не следует пытаться обрабатывать в собственной программе, поскольку они связаны с проблемами уровня <code>JVM</code>.
Например, исключения такого рода возникают, если закончилась память, доступная виртуальной машине.</p>
<p>В Java все исключения делятся на три типа: <strong>контролируемые исключения</strong> (<strong><code>checked</code></strong>)
и <strong>неконтролируемые исключения</strong> (<strong><code>unchecked</code></strong>), к которым относятся ошибки (<strong><code>Errors</code></strong>)
и исключения времени выполнения (<strong><code>RuntimeExceptions</code></strong>, потомок класса <strong><code>Exception</code></strong>).</p>
<p>Контролируемые исключения представляют собой ошибки, которые можно и нужно обрабатывать в программе,
к этому типу относятся все потомки класса <strong><code>Exception</code></strong> (кроме <strong><code>RuntimeException</code></strong>).</p>
<div class="code-element">
<div class="lang-line">
  <text>java</text>
  <button class="copy-button"
          id="code53bea4f98824957e8c7519132077182ab"
          onclick="copyCode(code53bea4f98824957e8c7519132077182a, code53bea4f98824957e8c7519132077182ab)">
    <svg stroke="currentColor"
         fill="none"
         stroke-width="2"
         viewBox="0 0 24 24"
         stroke-linecap="round"
         stroke-linejoin="round"
         class="h-4 w-4"
         height="1em"
         width="1em"
         xmlns="http://www.w3.org/2000/svg">
      <path d="M16 4h2a2 2 0 0 1 2 2v14a2 2 0 0 1-2 2H6a2 2 0 0 1-2-2V6a2 2 0 0 1 2-2h2"></path>
      <rect x="8" y="2" width="8" height="4" rx="1" ry="1"></rect>
    </svg>
    <text>Copy code</text>
  </button>

</div>
<div class="code" id="code53bea4f98824957e8c7519132077182a"><div class="highlight"><pre><span></span><span class="c1">// CustomException.java</span>
<span class="kd">public</span><span class="w"> </span><span class="kd">class</span> <span class="nc">CustomException</span><span class="w"> </span><span class="kd">extends</span><span class="w"> </span><span class="n">Exception</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="kd">public</span><span class="w"> </span><span class="nf">CustomException</span><span class="p">(</span><span class="n">String</span><span class="w"> </span><span class="n">message</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">        </span><span class="kd">super</span><span class="p">(</span><span class="n">message</span><span class="p">);</span>
<span class="w">    </span><span class="p">}</span>
<span class="p">}</span>

<span class="c1">// CustomExceptionExample.java</span>
<span class="kd">public</span><span class="w"> </span><span class="kd">class</span> <span class="nc">CustomExceptionExample</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="kd">public</span><span class="w"> </span><span class="kd">static</span><span class="w"> </span><span class="kt">void</span><span class="w"> </span><span class="nf">main</span><span class="p">(</span><span class="n">String</span><span class="o">[]</span><span class="w"> </span><span class="n">args</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">        </span><span class="k">try</span><span class="w"> </span><span class="p">{</span>
<span class="w">            </span><span class="k">throw</span><span class="w"> </span><span class="k">new</span><span class="w"> </span><span class="n">CustomException</span><span class="p">(</span><span class="s">&quot;Это собственное исключение!&quot;</span><span class="p">);</span>
<span class="w">        </span><span class="p">}</span><span class="w"> </span><span class="k">catch</span><span class="w"> </span><span class="p">(</span><span class="n">CustomException</span><span class="w"> </span><span class="n">e</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">            </span><span class="n">System</span><span class="p">.</span><span class="na">out</span><span class="p">.</span><span class="na">println</span><span class="p">(</span><span class="s">&quot;Поймано исключение: &quot;</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">e</span><span class="p">.</span><span class="na">getMessage</span><span class="p">());</span>
<span class="w">        </span><span class="p">}</span>
<span class="w">    </span><span class="p">}</span>
<span class="p">}</span>
</pre></div></div>
</div>